source/mir/math/stat.d

```diff
@@ -1626,6 +1627,704 @@ unittest
+            meanAccumulator.put(r.move);
+            sumOfSquares.put(r.move);
+        }
+        else
+        {
+            foreach(x; r)
+            {
+                meanAccumulator.put(x);
+                sumOfSquares.put(x);
+            }
+        }
+    }
+
+    ///
+    void put()(T x)
+    {
+        meanAccumulator.put(x);
+        sumOfSquares.put(x);
+    }
+    
+    ///
+    F variance(F = T)(bool isPopulation) @property
+    {
+        if (isPopulation == false)
+            return cast(F) sumOfSquares.sum / cast(F) (count - 1) - 
+                square(cast(F) meanAccumulator.mean) * (cast(F) count / cast(F) (count - 1));
+        else
+            return cast(F) sumOfSquares.sum / cast(F) count - 
+                square(cast(F) meanAccumulator.mean);
+    }
+}
+
+///
+version(mir_test)
+@safe pure nothrow
+unittest
+{
+    import mir.ndslice.slice: sliced;
+    import mir.math.common: approxEqual;
+
+    auto x = [0.0, 1.0, 1.5, 2.0, 3.5, 4.25,
+              2.0, 7.5, 5.0, 1.0, 1.5, 0.0].sliced;
+
+    enum PopulationTrueCT = true;
+    enum PopulationFalseCT = false;
+    bool PopulationTrueRT = true;
+    bool PopulationFalseRT = false;
+
+    VarianceAccumulator!(double, VarianceAlgo.naive, Summation.naive) v;
+    v.put(x);
+    assert(v.variance(PopulationTrueRT).approxEqual(54.76562 / 12));
+    assert(v.variance(PopulationTrueCT).approxEqual(54.76562 / 12));
+    assert(v.variance(PopulationFalseRT).approxEqual(54.76562 / 11));
+    assert(v.variance(PopulationFalseCT).approxEqual(54.76562 / 11));
+
+    v.put(4.0);
+    assert(v.variance(PopulationTrueRT).approxEqual(57.01923 / 13));
+    assert(v.variance(PopulationTrueCT).approxEqual(57.01923 / 13));
+    assert(v.variance(PopulationFalseRT).approxEqual(57.01923 / 12));
+    assert(v.variance(PopulationFalseCT).approxEqual(57.01923 / 12));
+}
+
+///
+struct VarianceAccumulator(T, VarianceAlgo varianceAlgo, Summation summation)
+    if (isMutable!T && 
+        varianceAlgo == VarianceAlgo.online)
+{
+    mixin variance_ops!(T, summation);
+
+    ///
+    MapSummator!(square, T, summation) centeredSumOfSquares;
+
+    ///
+    void put(Range)(Range r)
+        if (isIterable!Range)
+    {
+        import mir.ndslice.topology: vmap;
+        import mir.ndslice.internal: LeftOp;
+
+        if (count == 0) {
+            meanAccumulator.put(r.move);
+            centeredSumOfSquares.put(r.move.vmap(LeftOp!("-", T)(meanAccumulator.mean)));
+        } else {
+            T oldMean = meanAccumulator.mean;
+
+            meanAccumulator.put(r.move);
+            static if (hasShape!Range)
+            {
+                import mir.ndslice.topology: vmap, map, zip;
+                import mir.ndslice.internal: LeftOp;
+
+                centeredSumOfSquares.
+                    mapSummator.
+                    put(
+                        r.move.vmap(LeftOp!("-", T)(meanAccumulator.mean)).
+                        zip(r.move.vmap(LeftOp!("-", T)(oldMean))).
```

Comment:
twice `r.move` is incorrect. Only the second one argument can use `.move`. In practice it is better to define `auto c = r;` and then move `c` and `r` because D argument evaluation order seems to be broken.